LPOP/RPOP with count against non existing list return null array

enjoy-binbin opened the pull request:
It used to return `$-1` in RESP2, now we will return `*-1`.
This is a bug in redis 6.2 when COUNT was added, the `COUNT`
option was introduced in #8179. Fix #10089.

the documentation of [LPOP](https://redis.io/commands/lpop) says
```
When called without the count argument:
Bulk string reply: the value of the first element, or nil when key does not exist.

When called with the count argument:
Array reply: list of popped elements, or nil when key does not exist.
```

zuiderkwast commented on src/t_list.c:
```diff
@@ -501,7 +501,7 @@ void popGenericCommand(client *c, int where) {
     }
 
-    robj *o = lookupKeyWriteOrReply(c, c->argv[1], shared.null[c->resp]);
+    robj *o = lookupKeyWriteOrReply(c, c->argv[1], hascount ? shared.nullarray[c->resp]: shared.null[c->resp]);
```
I don't think we should complicate this. Users should not rely on which kind of null is returned.

The [protocol documentation](https://redis.io/topics/protocol) says:

> The concept of Null Array exists as well, and is an alternative way to specify a Null value (usually the Null Bulk String is used, but for historical reasons we have two formats).

oranagra replied:
@zuiderkwast i don't think this is a serious complication. the breaking change bothers me more (maybe the doc quote you posted reduces that).

considering the documentation we wrote for 6.2 (added in the top comment), the response should be a null array, and it seems more logical to fix the code rather than the doc.

zuiderkwast replied:
~~Should we document that it returns a null-bulk or null-array depending on the COUNT parameter? I think it's unnecessarily complex, not the code but the API.~~

~~Maybe it should be a null-array in all cases instead?~~

[Edit] Sorry about this. The fix makes sense considering the return is bulk vs multi-bulk depending on COUNT.

oranagra commented:
@enjoy-binbin can you scan other places where COUNT was added in 6.2 and other early exists of these commands to see if we don't have similar issues in other places.

enjoy-binbin commented:
sure, i can take a look tomorrow

madolson reviewed (approved):
+1 to verifying the other locations don't introduce breaking changes. I am in favor of fixing this so that it matches the 6.0 behavior. We should backport this to 6.2 as well.

oranagra commented:
@madolson it does match 6.0 already.
6.0 didn't have the count argument, so it always returned null string in this case

madolson commented:
@oranagra For some reason I thought this broke from 6.0->6.2, but I understand now, it doesn't change my opinion.

enjoy-binbin commented:
base on this branch (this PR)
> saying up front, i scanned most of the commands, here is the result:
> it looks like we've got everything covered

```c
shared.emptyarray = createObject(OBJ_STRING,sdsnew("*0\r\n"));

shared.null[2] = createObject(OBJ_STRING,sdsnew("$-1\r\n"));
shared.null[3] = createObject(OBJ_STRING,sdsnew("_\r\n"));

shared.nullarray[2] = createObject(OBJ_STRING,sdsnew("*-1\r\n"));
shared.nullarray[3] = createObject(OBJ_STRING,sdsnew("_\r\n"));

shared.emptymap[2] = createObject(OBJ_STRING,sdsnew("*0\r\n"));
shared.emptymap[3] = createObject(OBJ_STRING,sdsnew("%0\r\n"));

shared.emptyset[2] = createObject(OBJ_STRING,sdsnew("*0\r\n"));
shared.emptyset[3] = createObject(OBJ_STRING,sdsnew("~0\r\n"));
```


list:

LPOP/RPOP key [count]  (count added in 6.2)
- without count:
  - non existing: return shared.null[resp]
- with count:
  - non existing: now we return shared.nullarray[resp] (this PR fix it)
  - count is 0: return shared.emptyarray

-----------------------------------------------------------------------------

set:

SPOP key [count]  (count added in 3.2)
- without count:
  - non existing: return shared.null[c->resp]
- with count:
  - non existing: return shared.emptyset[c->resp] (used to return shared.null[resp], we fixed it in #4709)
  - count is 0: return shared.emptyset[resp] (used to return addReplyNull, ie: shared.null[resp], fixed it in 06d4903)


SRANDMEMBER
- without count:
  - non existing: return shared.null[c->resp]
- with count:
  - non existing key / count is 0: return shared.emptyarray (used to return emptyset[c->resp], fixed it in #8504)


-----------------------------------------------------------------------------


zset:

ZPOPMIN/ZPOPMAX key [count]
- without count:
  - non existing zset: return shared.emptyarray
- with count:
  - non existing: return shared.emptyarray
  - count is 0: return shared.emptyarray


ZRANDMEMBER
- without count: 
  - non existing: return shared.null[c->resp]
- with count:
  - non existing: return shared.emptyarray (used to return null[c->resp], fixed it in #9178)
  - count is 0: return shared.emptyarray


-----------------------------------------------------------------------------


hash:

HGETALL key
non existing: return shared.emptymap[c->resp] (used to return shared.emptyarray, fixed it in #7781)

HRANDFIELD
- without count: 
  - non existing: return shared.null[c->resp]
- with count:
  - non existing: return shared.emptyarray (used to return null[c->resp], fixed it in #9178)
  - count is 0: return shared.emptyarray



-----------------------------------------------------------------------------

geo:

georadius*
geosearch
- non existing key: return addReply(c, shared.emptyarray);
- non existing key with store key: return 0 (used to return emptyarray, fixed it in #9271)
